L13-python-interceptors.md

```diff
@@ -0,0 +1,274 @@
+
+1. Wraps over low-level abstraction like startBatch, Operations (Java, Node.js).
+2. Wraps over high-level abstraction like Server, Channel (Python, Golang).
+
+The interceptor API of the first approach enables gRPC developers to add handlers to the various event, take [Node.js Interceptor design](https://github.com/grpc/proposal/blob/master/L5-node-client-interceptors.md) as an example:
+
+1. __start__: An interception method called before an outbound call has started.
+2. __sendMessage__: An interception method called prior to every outbound message.
+3. __halfClose__: An interception method called when the outbound stream closes (after the message is sent).
+4. __cancel__: An interception method called when the stream is canceled from the client.
+5. __onReceiveMetadata__: An inbound interception method triggered when metadata is received.
+6. __onReceiveMessage__: An inbound interception method triggered when each message is received.
+7. __onReceiveStatus__: An inbound interception method triggered when status is received.
+
+The interceptor API of the second approach requires developers to wrap logic over the invocation of the RPC call, take [Python unary-unary Interceptor](https://grpc.io/grpc/python/grpc.html#client-side-interceptor) as an example:
+
+```Python
+class _ExampleClientInterceptor(grpc.UnaryUnaryClientInterceptor):
+
+    def intercept_unary_unary(self, continuation, client_call_details, request):
+        new_details, new_request = <Pre Process Logic>
+
+        response = continuation(new_details, new_request)
+
+        new_response = <Post Process Logic>
+        return new_response  # It's both a grpc.Call and a grpc.Future.
+
+```
+
+The difference between mehrdada@'s gRFC and implementation is that: 1) the function signature changed; 2) the actual implementation doesn't support asynchronous interception; 3) the concept of `invoker` is replaced by `continuation`; 4) the actual implementation packs credentials/timeout/wait_for_ready/method/metadata into `grpc.ClientCallDetails`.
+
+### Server-Side Interceptor Semantic Comparison
+
+[Java server interceptor](https://grpc.io/grpc-java/javadoc/) allows developers to register event handlers:
+
+1. **onCancel**: The call was cancelled and the server is encouraged to abort processing to save resources, since the client will not process any further messages.
+2. **onComplete**: The call is considered complete and onCancel() is guaranteed not to be called.
+3. **onHalfClose**: The client completed all message sending.
+4. **onMessage**: A request message has been received.
+5. **onReady**: This indicates that the call is now capable of sending additional messages without requiring excessive buffering internally.
+
+Golang server interceptor and Python server interceptor **by design** will achieve the similar functionality of the client interceptor that the interception logic is handled inside one function. Here is the definition of Golang's unary server interceptor:
```

Comment:
s/the similar functionality of/similar functionality as/

s/interceptor that/interceptor, where/